src/lib/utils/application-commands/ApplicationCommandRegistry.ts

```diff
@@ -0,0 +1,465 @@
+			if (!guildCommands) {
+				this.debug(`There are no commands for guild with id "${guildId}". Will create ${type} command "${commandName}".`);
+				await this.createMissingCommand(commandsManager, builtData, type, guildId);
+				continue;
+			}
+
+			const existingGuildCommand = guildCommands.find(findCallback);
+
+			if (existingGuildCommand) {
+				this.debug(`Checking if guild ${type} command "${commandName}" is identical to command "${existingGuildCommand.id}"`);
+
+				switch (apiCall.type) {
+					case InternalRegistryAPIType.ChatInput:
+						this.addChatInputCommandIds(existingGuildCommand.id);
+						break;
+					case InternalRegistryAPIType.ContextMenu:
+						this.addContextMenuCommandIds(existingGuildCommand.id);
+						break;
+				}
+
+				await this.handleCommandPresent(existingGuildCommand, builtData, behaviorIfNotEqual);
+			} else if (registerOptions.registerCommandIfMissing ?? true) {
+				this.debug(`Creating new guild ${type} command with name "${commandName}" for guild "${guildId}"`);
+				await this.createMissingCommand(commandsManager, builtData, type, guildId);
+			} else {
+				this.debug(`Doing nothing about missing guild ${type} command with name "${commandName}" for guild "${guildId}"`);
+			}
+		}
+	}
+
+	private async handleCommandPresent(
+		applicationCommand: ApplicationCommand,
+		apiData: InternalAPICall['builtData'],
+		behaviorIfNotEqual: RegisterBehavior,
+		guildId?: string
+	) {
+		const now = Date.now();
+
+		// Step 0: compute differences
+		const differences = getCommandDifferences(convertApplicationCommandToApiData(applicationCommand), apiData);
+
+		const later = Date.now() - now;
+		this.debug(`Took ${later}ms to process differences`);
+
+		// Step 1: if there are no differences, return
+		if (!differences.length) {
+			this.debug(
+				`${guildId ? 'Guild command' : 'Command'} "${apiData.name}" is identical to command "${applicationCommand.name}" (${
+					applicationCommand.id
+				})`
+			);
+			return;
+		}
+
+		this.logCommandDifferences(differences, applicationCommand, behaviorIfNotEqual === RegisterBehavior.LogToConsole);
+
+		// Step 2: if the behavior is to log to console, log the differences
+		if (behaviorIfNotEqual === RegisterBehavior.LogToConsole) {
+			return;
+		}
+
+		// Step 3: if the behavior is to update, update the command
+		try {
+			await applicationCommand.edit(apiData as ChatInputApplicationCommandData);
+			this.debug(`Updated command ${applicationCommand.name} (${applicationCommand.id}) with new api data`);
+		} catch (error) {
+			this.error(`Failed to update command ${applicationCommand.name} (${applicationCommand.id})`, error);
+		}
+	}
+
+	private logCommandDifferences(differences: CommandDifference[], applicationCommand: ApplicationCommand, logAsWarn: boolean) {
+		const finalMessage: string[] = [];
+
+		for (const difference of differences) {
+			finalMessage.push(
+				[
+					`└── At path: ${difference.key}`, //
+					`     ├── Received: ${difference.original}`,
```

Comment:
Do we trust this indenting to always go right? Wouldn't it be better to use
```ts
const pad = ' '.repeat(5);

// ...
`${pad}└── Expected: ${difference.expected}
```

Reply:
Might be a better idea, will do